File: rs/moq-lite/src/ietf/parameters.rs
use std::collections::{BTreeMap, HashMap, btree_map, hash_map};

use num_enum::{FromPrimitive, IntoPrimitive};

use crate::coding::*;

use super::Version;

const MAX_PARAMS: u64 = 64;
const PARAM_SUBVALUE_VERSION: Version = Version::Draft15;

// ---- Setup Parameters (used in CLIENT_SETUP/SERVER_SETUP) ----

#[derive(Debug, Copy, Clone, FromPrimitive, IntoPrimitive, Eq, Hash, PartialEq)]
#[repr(u64)]
pub enum ParameterVarInt {
	MaxRequestId = 2,
	MaxAuthTokenCacheSize = 4,
	#[num_enum(catch_all)]
	Unknown(u64),
}

#[derive(Debug, Copy, Clone, FromPrimitive, IntoPrimitive, Eq, Hash, PartialEq)]
#[repr(u64)]
pub enum ParameterBytes {
	Path = 1,
	AuthorizationToken = 3,
	Authority = 5,
	Implementation = 7,
	#[num_enum(catch_all)]
	Unknown(u64),
}

#[derive(Default, Debug, Clone)]
pub struct Parameters {
	vars: HashMap<ParameterVarInt, u64>,
	bytes: HashMap<ParameterBytes, Vec<u8>>,
}

impl Decode<Version> for Parameters {
	fn decode<R: bytes::Buf>(mut r: &mut R, version: Version) -> Result<Self, DecodeError> {
		let mut vars = HashMap::new();
		let mut bytes = HashMap::new();

		let count = u64::decode(r, version)?;

		if count > MAX_PARAMS {
			return Err(DecodeError::TooMany);
		}

		let mut prev_type: u64 = 0;

		for i in 0..count {
			let kind = match version {
				Version::Draft16 => {
					let delta = u64::decode(r, version)?;
					let abs = if i == 0 { delta } else { prev_type + delta };
					prev_type = abs;
					abs
				}
				Version::Draft14 | Version::Draft15 | Version::Draft17 => u64::decode(r, version)?,
			};

			if kind % 2 == 0 {
				let kind = ParameterVarInt::from(kind);
				match vars.entry(kind) {
					hash_map::Entry::Occupied(_) => return Err(DecodeError::Duplicate),
					hash_map::Entry::Vacant(entry) => entry.insert(u64::decode(&mut r, version)?),
				};
			} else {
				let kind = ParameterBytes::from(kind);
				match bytes.entry(kind) {
					hash_map::Entry::Occupied(_) => return Err(DecodeError::Duplicate),
					hash_map::Entry::Vacant(entry) => entry.insert(Vec::<u8>::decode(&mut r, version)?),
				};
			}
		}

		Ok(Parameters { vars, bytes })
	}
}

impl Encode<Version> for Parameters {
	fn encode<W: bytes::BufMut>(&self, w: &mut W, version: Version) -> Result<(), EncodeError> {
		let count = self.vars.len() + self.bytes.len();
		if count as u64 > MAX_PARAMS {
			return Err(EncodeError::TooMany);
		}
		count.encode(w, version)?;

		match version {
			Version::Draft16 => {
				// Delta encoding: collect all keys, sort, encode deltas
				let mut all: Vec<(u64, bool, usize)> = Vec::new(); // (key, is_var, index)
				let var_keys: Vec<_> = self.vars.keys().collect();
				let byte_keys: Vec<_> = self.bytes.keys().collect();
				for (i, k) in var_keys.iter().enumerate() {
					all.push((u64::from(**k), true, i));
				}
				for (i, k) in byte_keys.iter().enumerate() {
					all.push((u64::from(**k), false, i));
				}
				all.sort_by_key(|(k, _, _)| *k);

				let var_vals: Vec<_> = self.vars.values().collect();
				let byte_vals: Vec<_> = self.bytes.values().collect();

				let mut prev_type: u64 = 0;
				for (idx, (kind, is_var, orig_idx)) in all.iter().enumerate() {
					let delta = if idx == 0 { *kind } else { kind - prev_type };
					prev_type = *kind;
					delta.encode(w, version)?;

					if *is_var {
						var_vals[*orig_idx].encode(w, version)?;
					} else {
						byte_vals[*orig_idx].encode(w, version)?;
					}
				}
			}
			Version::Draft14 | Version::Draft15 | Version::Draft17 => {
				for (kind, value) in self.vars.iter() {
					u64::from(*kind).encode(w, version)?;
					value.encode(w, version)?;
				}

				for (kind, value) in self.bytes.iter() {
					u64::from(*kind).encode(w, version)?;
					value.encode(w, version)?;
				}
			}
		}

		Ok(())
	}
}

impl Parameters {
	pub fn get_varint(&self, kind: ParameterVarInt) -> Option<u64> {
		self.vars.get(&kind).copied()
	}

	pub fn set_varint(&mut self, kind: ParameterVarInt, value: u64) {
		self.vars.insert(kind, value);
	}

	#[cfg(test)]
	pub fn get_bytes(&self, kind: ParameterBytes) -> Option<&[u8]> {
		self.bytes.get(&kind).map(|v| v.as_slice())
	}

	pub fn set_bytes(&mut self, kind: ParameterBytes, value: Vec<u8>) {
		self.bytes.insert(kind, value);
	}
}

// ---- Message Parameters (used in Subscribe, Publish, Fetch, etc.) ----
// Uses raw u64 keys since parameter IDs have different meanings from setup parameters.
// BTreeMap ensures deterministic wire encoding order.

#[derive(Default, Debug, Clone)]
pub struct MessageParameters {
	vars: BTreeMap<u64, u64>,
	bytes: BTreeMap<u64, Vec<u8>>,
}

impl Decode<Version> for MessageParameters {
	fn decode<R: bytes::Buf>(mut r: &mut R, version: Version) -> Result<Self, DecodeError> {
		let mut vars = BTreeMap::new();
		let mut bytes = BTreeMap::new();

		let count = u64::decode(r, version)?;

		if count > MAX_PARAMS {
			return Err(DecodeError::TooMany);
		}

		let mut prev_type: u64 = 0;

		for i in 0..count {
			let kind = match version {
				Version::Draft16 => {
					let delta = u64::decode(r, version)?;
					let abs = if i == 0 { delta } else { prev_type + delta };
					prev_type = abs;
					abs
				}
				Version::Draft14 | Version::Draft15 | Version::Draft17 => u64::decode(r, version)?,
			};

			if kind % 2 == 0 {
				match vars.entry(kind) {
					btree_map::Entry::Occupied(_) => return Err(DecodeError::Duplicate),
					btree_map::Entry::Vacant(entry) => entry.insert(u64::decode(&mut r, version)?),
				};
			} else {
				match bytes.entry(kind) {
					btree_map::Entry::Occupied(_) => return Err(DecodeError::Duplicate),
					btree_map::Entry::Vacant(entry) => entry.insert(Vec::<u8>::decode(&mut r, version)?),
				};
			}
		}

		Ok(MessageParameters { vars, bytes })
	}
}

impl Encode<Version> for MessageParameters {
	fn encode<W: bytes::BufMut>(&self, w: &mut W, version: Version) -> Result<(), EncodeError> {
		let count = self.vars.len() + self.bytes.len();
		if count as u64 > MAX_PARAMS {
			return Err(EncodeError::TooMany);
		}
		count.encode(w, version)?;

		match version {
			Version::Draft16 => {
				// Delta encoding: BTreeMap is already sorted, merge and sort by key
				enum ParamValue<'a> {
					Var(&'a u64),
					Bytes(&'a Vec<u8>),
				}
				let mut all: Vec<(u64, ParamValue)> = Vec::new();
				for (k, v) in self.vars.iter() {
					all.push((*k, ParamValue::Var(v)));
				}
				for (k, v) in self.bytes.iter() {
					all.push((*k, ParamValue::Bytes(v)));
				}
				all.sort_by_key(|(k, _)| *k);

				let mut prev_type: u64 = 0;
				for (idx, (kind, val)) in all.iter().enumerate() {
					let delta = if idx == 0 { *kind } else { kind - prev_type };
					prev_type = *kind;
					delta.encode(w, version)?;

					match val {
						ParamValue::Var(v) => v.encode(w, version)?,
						ParamValue::Bytes(v) => v.encode(w, version)?,
					}
				}
			}
			Version::Draft14 | Version::Draft15 | Version::Draft17 => {
				for (kind, value) in self.vars.iter() {
					kind.encode(w, version)?;
					value.encode(w, version)?;
				}

				for (kind, value) in self.bytes.iter() {
					kind.encode(w, version)?;
					value.encode(w, version)?;
				}
			}
		}

		Ok(())
	}
}

impl MessageParameters {
	// Varint parameter IDs (even)
	//const DELIVERY_TIMEOUT: u64 = 0x02;
	//const MAX_CACHE_DURATION: u64 = 0x04;
	//const EXPIRES: u64 = 0x08;
	//const PUBLISHER_PRIORITY: u64 = 0x0E;
	const FORWARD: u64 = 0x10;
	const SUBSCRIBER_PRIORITY: u64 = 0x20;
	const GROUP_ORDER: u64 = 0x22;

	// Bytes parameter IDs (odd)
	#[allow(dead_code)]
	const AUTHORIZATION_TOKEN: u64 = 0x03;
	const LARGEST_OBJECT: u64 = 0x09;
	const SUBSCRIPTION_FILTER: u64 = 0x21;

	// --- Varint accessors ---

	/*
	pub fn delivery_timeout(&self) -> Option<u64> {
		self.vars.get(&Self::DELIVERY_TIMEOUT).copied()
	}

	pub fn set_delivery_timeout(&mut self, v: u64) {
		self.vars.insert(Self::DELIVERY_TIMEOUT, v);
	}

	pub fn max_cache_duration(&self) -> Option<u64> {
		self.vars.get(&Self::MAX_CACHE_DURATION).copied()
	}

	pub fn set_max_cache_duration(&mut self, v: u64) {
		self.vars.insert(Self::MAX_CACHE_DURATION, v);
	}

	pub fn expires(&self) -> Option<u64> {
		self.vars.get(&Self::EXPIRES).copied()
	}

	pub fn set_expires(&mut self, v: u64) {
		self.vars.insert(Self::EXPIRES, v);
	}

	pub fn publisher_priority(&self) -> Option<u8> {
		self.vars.get(&Self::PUBLISHER_PRIORITY).map(|v| *v as u8)
	}

	pub fn set_publisher_priority(&mut self, v: u8) {
		self.vars.insert(Self::PUBLISHER_PRIORITY, v as u64);
	}
	*/

	pub fn forward(&self) -> Option<bool> {
		self.vars.get(&Self::FORWARD).map(|v| *v != 0)
	}

	pub fn set_forward(&mut self, v: bool) {
		self.vars.insert(Self::FORWARD, v as u64);
	}

	pub fn subscriber_priority(&self) -> Option<u8> {
		self.vars.get(&Self::SUBSCRIBER_PRIORITY).map(|v| *v as u8)
	}

	pub fn set_subscriber_priority(&mut self, v: u8) {
		self.vars.insert(Self::SUBSCRIBER_PRIORITY, v as u64);
	}

	pub fn group_order(&self) -> Option<u64> {
		self.vars.get(&Self::GROUP_ORDER).copied()
	}

	pub fn set_group_order(&mut self, v: u64) {
		self.vars.insert(Self::GROUP_ORDER, v);
	}

	// --- Bytes accessors ---

	/// Get largest object location (encoded as group_id varint + object_id varint)
	pub fn largest_object(&self) -> Option<super::Location> {
		let data = self.bytes.get(&Self::LARGEST_OBJECT)?;
		let mut buf = bytes::Bytes::from(data.clone());
		// Sub-values within parameters always use QUIC varint encoding.
		let v = PARAM_SUBVALUE_VERSION;
		let group = u64::decode(&mut buf, v).ok()?;
		let object = u64::decode(&mut buf, v).ok()?;
		Some(super::Location { group, object })
	}

	pub fn set_largest_object(&mut self, loc: &super::Location) -> Result<(), EncodeError> {
		let mut buf = Vec::new();
		// Sub-values within parameters always use QUIC varint encoding.
		let v = PARAM_SUBVALUE_VERSION;
		loc.group.encode(&mut buf, v)?;
		loc.object.encode(&mut buf, v)?;
		self.bytes.insert(Self::LARGEST_OBJECT, buf);
		Ok(())
	}

	/// Get subscription filter (encoded as filter_type varint [+ filter data])
	pub fn subscription_filter(&self) -> Option<super::FilterType> {
		let data = self.bytes.get(&Self::SUBSCRIPTION_FILTER)?;
		let mut buf = bytes::Bytes::from(data.clone());
		// Sub-values within parameters always use QUIC varint encoding.
		super::FilterType::decode(&mut buf, PARAM_SUBVALUE_VERSION).ok()
	}

	pub fn set_subscription_filter(&mut self, ft: super::FilterType) -> Result<(), EncodeError> {
		let mut buf = Vec::new();
		// Sub-values within parameters always use QUIC varint encoding.
		ft.encode(&mut buf, PARAM_SUBVALUE_VERSION)?;
		self.bytes.insert(Self::SUBSCRIPTION_FILTER, buf);
		Ok(())
	}
}

#[cfg(test)]
mod tests {
	use super::*;
	use bytes::BytesMut;

	#[test]
	fn test_parameters_v16_delta_round_trip() {
		let mut params = Parameters::default();
		params.set_bytes(ParameterBytes::Path, b"/test".to_vec());
		params.set_varint(ParameterVarInt::MaxRequestId, 100);
		params.set_bytes(ParameterBytes::Implementation, b"test-impl".to_vec());

		let mut buf = BytesMut::new();
		params.encode(&mut buf, Version::Draft16).unwrap();

		let mut bytes = buf.freeze();
		let decoded = Parameters::decode(&mut bytes, Version::Draft16).unwrap();

		assert_eq!(decoded.get_bytes(ParameterBytes::Path), Some(b"/test".as_ref()));
		assert_eq!(decoded.get_varint(ParameterVarInt::MaxRequestId), Some(100));
		assert_eq!(
			decoded.get_bytes(ParameterBytes::Implementation),
			Some(b"test-impl".as_ref())
		);
	}

	#[test]
	fn test_parameters_v15_round_trip() {
		let mut params = Parameters::default();
		params.set_bytes(ParameterBytes::Path, b"/test".to_vec());
		params.set_varint(ParameterVarInt::MaxRequestId, 100);

		let mut buf = BytesMut::new();
		params.encode(&mut buf, Version::Draft15).unwrap();

		let mut bytes = buf.freeze();
		let decoded = Parameters::decode(&mut bytes, Version::Draft15).unwrap();

		assert_eq!(decoded.get_bytes(ParameterBytes::Path), Some(b"/test".as_ref()));
		assert_eq!(decoded.get_varint(ParameterVarInt::MaxRequestId), Some(100));
	}

	#[test]
	fn test_message_parameters_v16_delta_round_trip() {
		let mut params = MessageParameters::default();
		params.set_subscriber_priority(200);
		params.set_group_order(2);
		params.set_forward(true);

		let mut buf = BytesMut::new();
		params.encode(&mut buf, Version::Draft16).unwrap();

		let mut bytes = buf.freeze();
		let decoded = MessageParameters::decode(&mut bytes, Version::Draft16).unwrap();

		assert_eq!(decoded.subscriber_priority(), Some(200));
		assert_eq!(decoded.group_order(), Some(2));
		assert_eq!(decoded.forward(), Some(true));
	}

	#[test]
	fn test_message_parameters_v15_round_trip() {
		let mut params = MessageParameters::default();
		params.set_subscriber_priority(128);
		params.set_group_order(2);

		let mut buf = BytesMut::new();
		params.encode(&mut buf, Version::Draft15).unwrap();

		let mut bytes = buf.freeze();
		let decoded = MessageParameters::decode(&mut bytes, Version::Draft15).unwrap();

		assert_eq!(decoded.subscriber_priority(), Some(128));
		assert_eq!(decoded.group_order(), Some(2));
	}

	#[test]
	fn test_message_parameters_v17_round_trip() {
		use crate::ietf::{FilterType, Location};

		let mut params = MessageParameters::default();
		params.set_subscriber_priority(200);
		params.set_group_order(2);
		params.set_forward(true);
		params.set_largest_object(&Location { group: 5, object: 3 }).unwrap();
		params.set_subscription_filter(FilterType::LargestObject).unwrap();

		let mut buf = BytesMut::new();
		params.encode(&mut buf, Version::Draft17).unwrap();

		let mut bytes = buf.freeze();
		let decoded = MessageParameters::decode(&mut bytes, Version::Draft17).unwrap();

		assert_eq!(decoded.subscriber_priority(), Some(200));
		assert_eq!(decoded.group_order(), Some(2));
		assert_eq!(decoded.forward(), Some(true));
		assert_eq!(decoded.largest_object(), Some(Location { group: 5, object: 3 }));
		assert_eq!(decoded.subscription_filter(), Some(FilterType::LargestObject));
	}

	#[test]
	fn test_message_parameters_empty_v16() {
		let params = MessageParameters::default();

		let mut buf = BytesMut::new();
		params.encode(&mut buf, Version::Draft16).unwrap();

		let mut bytes = buf.freeze();
		let decoded = MessageParameters::decode(&mut bytes, Version::Draft16).unwrap();

		assert_eq!(decoded.subscriber_priority(), None);
	}
}

File: js/lite/src/ietf/parameters.ts
import type { Reader, Writer } from "../stream.ts";
import * as Varint from "../varint.ts";
import { type IetfVersion, Version } from "./version.ts";

/// Setup Option key constants (separate namespace from Message Parameters).
export const SetupOption = {
	Path: 1n,
	MaxRequestId: 2n,
	AuthorizationToken: 3n,
	MaxAuthTokenCacheSize: 4n,
	Authority: 5n,
	Implementation: 7n,
} as const;

/// Setup Options — used in SETUP messages.
///
/// In d14-d16 these are count-prefixed ("Setup Parameters").
/// In d17 these have no count prefix ("Setup Options") and read/write to end of message.
export class SetupOptions {
	vars: Map<bigint, bigint>;
	bytes: Map<bigint, Uint8Array>;

	constructor() {
		this.vars = new Map();
		this.bytes = new Map();
	}

	get size() {
		return this.vars.size + this.bytes.size;
	}

	setBytes(id: bigint, value: Uint8Array) {
		if (id % 2n !== 1n) {
			throw new Error(`invalid parameter id: ${id.toString()}, must be odd`);
		}
		this.bytes.set(id, value);
	}

	setVarint(id: bigint, value: bigint) {
		if (id % 2n !== 0n) {
			throw new Error(`invalid parameter id: ${id.toString()}, must be even`);
		}
		this.vars.set(id, value);
	}

	getBytes(id: bigint): Uint8Array | undefined {
		if (id % 2n !== 1n) {
			throw new Error(`invalid parameter id: ${id.toString()}, must be odd`);
		}
		return this.bytes.get(id);
	}

	getVarint(id: bigint): bigint | undefined {
		if (id % 2n !== 0n) {
			throw new Error(`invalid parameter id: ${id.toString()}, must be even`);
		}
		return this.vars.get(id);
	}

	removeBytes(id: bigint): boolean {
		if (id % 2n !== 1n) {
			throw new Error(`invalid parameter id: ${id.toString()}, must be odd`);
		}
		return this.bytes.delete(id);
	}

	removeVarint(id: bigint): boolean {
		if (id % 2n !== 0n) {
			throw new Error(`invalid parameter id: ${id.toString()}, must be even`);
		}
		return this.vars.delete(id);
	}

	async encode(w: Writer, version: IetfVersion) {
		if (version === Version.DRAFT_16 || version === Version.DRAFT_17) {
			// d17: no count prefix; d16: count prefix
			if (version !== Version.DRAFT_17) {
				await w.u53(this.vars.size + this.bytes.size);
			}

			// Delta encoding: collect all keys, sort, encode deltas
			const all: { key: bigint; isVar: boolean }[] = [];
			for (const id of this.vars.keys()) all.push({ key: id, isVar: true });
			for (const id of this.bytes.keys()) all.push({ key: id, isVar: false });
			all.sort((a, b) => (a.key < b.key ? -1 : a.key > b.key ? 1 : 0));

			let prevId = 0n;
			for (let i = 0; i < all.length; i++) {
				const { key, isVar } = all[i];
				const delta = i === 0 ? key : key - prevId;
				prevId = key;
				await w.u62(delta);

				if (isVar) {
					// biome-ignore lint/style/noNonNullAssertion: key is guaranteed to exist in vars map
					await w.u62(this.vars.get(key)!);
				} else {
					// biome-ignore lint/style/noNonNullAssertion: key is guaranteed to exist in bytes map
					const value = this.bytes.get(key)!;
					await w.u53(value.length);
					await w.write(value);
				}
			}
		} else {
			await w.u53(this.vars.size + this.bytes.size);

			for (const [id, value] of this.vars) {
				await w.u62(id);
				await w.u62(value);
			}

			for (const [id, value] of this.bytes) {
				await w.u62(id);
				await w.u53(value.length);
				await w.write(value);
			}
		}
	}

	static async decode(r: Reader, version: IetfVersion): Promise<SetupOptions> {
		const params = new SetupOptions();

		if (version === Version.DRAFT_17) {
			// d17: no count prefix, read until reader is done
			let prevType = 0n;
			let i = 0;
			while (!(await r.done())) {
				const delta = await r.u62();
				const id = i === 0 ? delta : prevType + delta;
				prevType = id;
				i++;

				if (id % 2n === 0n) {
					if (params.vars.has(id)) {
						throw new Error(`duplicate parameter id: ${id.toString()}`);
					}
					const varint = await r.u62();
					params.setVarint(id, varint);
				} else {
					if (params.bytes.has(id)) {
						throw new Error(`duplicate parameter id: ${id.toString()}`);
					}
					const size = await r.u53();
					const bytes = await r.read(size);
					params.setBytes(id, bytes);
				}
			}
		} else {
			const count = await r.u53();
			let prevType = 0n;

			for (let i = 0; i < count; i++) {
				let id: bigint;
				if (version === Version.DRAFT_16) {
					const delta = await r.u62();
					id = i === 0 ? delta : prevType + delta;
					prevType = id;
				} else {
					id = await r.u62();
				}

				if (id % 2n === 0n) {
					if (params.vars.has(id)) {
						throw new Error(`duplicate parameter id: ${id.toString()}`);
					}
					const varint = await r.u62();
					params.setVarint(id, varint);
				} else {
					if (params.bytes.has(id)) {
						throw new Error(`duplicate parameter id: ${id.toString()}`);
					}
					const size = await r.u53();
					const bytes = await r.read(size);
					params.setBytes(id, bytes);
				}
			}
		}

		return params;
	}
}

// ---- Message Parameters (used in Subscribe, Publish, Fetch, etc.) ----
// Count-prefixed KVPs with delta-encoded keys (d16+).

// Varint parameter IDs (even)
const MSG_PARAM_DELIVERY_TIMEOUT = 0x02n;
const MSG_PARAM_MAX_CACHE_DURATION = 0x04n;
const MSG_PARAM_EXPIRES = 0x08n;
const MSG_PARAM_PUBLISHER_PRIORITY = 0x0en;
const MSG_PARAM_FORWARD = 0x10n;
const MSG_PARAM_SUBSCRIBER_PRIORITY = 0x20n;
const MSG_PARAM_GROUP_ORDER = 0x22n;

// Bytes parameter IDs (odd)
const MSG_PARAM_LARGEST_OBJECT = 0x09n;
const MSG_PARAM_SUBSCRIPTION_FILTER = 0x21n;

/// Message Parameters — count-prefixed KVPs used in control messages.
export class Parameters {
	vars: Map<bigint, bigint>;
	bytes: Map<bigint, Uint8Array>;

	constructor() {
		this.vars = new Map();
		this.bytes = new Map();
	}

	// --- Varint accessors ---

	get subscriberPriority(): number | undefined {
		const v = this.vars.get(MSG_PARAM_SUBSCRIBER_PRIORITY);
		return v !== undefined ? Number(v) : undefined;
	}

	set subscriberPriority(v: number) {
		this.vars.set(MSG_PARAM_SUBSCRIBER_PRIORITY, BigInt(v));
	}

	get groupOrder(): number | undefined {
		const v = this.vars.get(MSG_PARAM_GROUP_ORDER);
		return v !== undefined ? Number(v) : undefined;
	}

	set groupOrder(v: number) {
		this.vars.set(MSG_PARAM_GROUP_ORDER, BigInt(v));
	}

	get forward(): boolean | undefined {
		const v = this.vars.get(MSG_PARAM_FORWARD);
		return v !== undefined ? v !== 0n : undefined;
	}

	set forward(v: boolean) {
		this.vars.set(MSG_PARAM_FORWARD, v ? 1n : 0n);
	}

	get publisherPriority(): number | undefined {
		const v = this.vars.get(MSG_PARAM_PUBLISHER_PRIORITY);
		return v !== undefined ? Number(v) : undefined;
	}

	set publisherPriority(v: number) {
		this.vars.set(MSG_PARAM_PUBLISHER_PRIORITY, BigInt(v));
	}

	get expires(): bigint | undefined {
		return this.vars.get(MSG_PARAM_EXPIRES);
	}

	set expires(v: bigint) {
		this.vars.set(MSG_PARAM_EXPIRES, v);
	}

	get deliveryTimeout(): bigint | undefined {
		return this.vars.get(MSG_PARAM_DELIVERY_TIMEOUT);
	}

	set deliveryTimeout(v: bigint) {
		this.vars.set(MSG_PARAM_DELIVERY_TIMEOUT, v);
	}

	get maxCacheDuration(): bigint | undefined {
		return this.vars.get(MSG_PARAM_MAX_CACHE_DURATION);
	}

	set maxCacheDuration(v: bigint) {
		this.vars.set(MSG_PARAM_MAX_CACHE_DURATION, v);
	}

	// --- Bytes accessors ---

	get largest(): { groupId: bigint; objectId: bigint } | undefined {
		const data = this.bytes.get(MSG_PARAM_LARGEST_OBJECT);
		if (!data || data.length === 0) return undefined;
		const [groupId, rest] = Varint.decode(data);
		const [objectId] = Varint.decode(rest);
		return { groupId: BigInt(groupId), objectId: BigInt(objectId) };
	}

	set largest(v: { groupId: bigint; objectId: bigint }) {
		const buf1 = Varint.encode(Number(v.groupId));
		const buf2 = Varint.encode(Number(v.objectId));
		const combined = new Uint8Array(buf1.length + buf2.length);
		combined.set(buf1, 0);
		combined.set(buf2, buf1.length);
		this.bytes.set(MSG_PARAM_LARGEST_OBJECT, combined);
	}

	get subscriptionFilter(): number | undefined {
		const data = this.bytes.get(MSG_PARAM_SUBSCRIPTION_FILTER);
		if (!data || data.length === 0) return undefined;
		// Filter type is a varint — for our purposes, the first byte suffices
		return data[0];
	}

	set subscriptionFilter(v: number) {
		this.bytes.set(MSG_PARAM_SUBSCRIPTION_FILTER, new Uint8Array([v]));
	}

	async encode(w: Writer, version: IetfVersion) {
		await w.u53(this.vars.size + this.bytes.size);

		if (version === Version.DRAFT_14 || version === Version.DRAFT_15) {
			for (const [id, value] of this.vars) {
				await w.u62(id);
				await w.u62(value);
			}

			for (const [id, value] of this.bytes) {
				await w.u62(id);
				await w.u53(value.length);
				await w.write(value);
			}
		} else {
			// d16+: Delta encoding, merge vars and bytes, sort by key
			const all: { key: bigint; isVar: boolean }[] = [];
			for (const id of this.vars.keys()) all.push({ key: id, isVar: true });
			for (const id of this.bytes.keys()) all.push({ key: id, isVar: false });
			all.sort((a, b) => (a.key < b.key ? -1 : a.key > b.key ? 1 : 0));

			let prevId = 0n;
			for (let i = 0; i < all.length; i++) {
				const { key, isVar } = all[i];
				const delta = i === 0 ? key : key - prevId;
				prevId = key;
				await w.u62(delta);

				if (isVar) {
					// biome-ignore lint/style/noNonNullAssertion: key is guaranteed to exist in vars map
					await w.u62(this.vars.get(key)!);
				} else {
					// biome-ignore lint/style/noNonNullAssertion: key is guaranteed to exist in bytes map
					const value = this.bytes.get(key)!;
					await w.u53(value.length);
					await w.write(value);
				}
			}
		}
	}

	static async decode(r: Reader, version: IetfVersion): Promise<Parameters> {
		const count = await r.u53();
		const params = new Parameters();

		let prevType = 0n;

		for (let i = 0; i < count; i++) {
			let id: bigint;
			if (version === Version.DRAFT_14 || version === Version.DRAFT_15) {
				id = await r.u62();
			} else {
				// d16+: delta encoding
				const delta = await r.u62();
				id = i === 0 ? delta : prevType + delta;
				prevType = id;
			}

			if (id % 2n === 0n) {
				if (params.vars.has(id)) {
					throw new Error(`duplicate message parameter id: ${id.toString()}`);
				}
				const varint = await r.u62();
				params.vars.set(id, varint);
			} else {
				if (params.bytes.has(id)) {
					throw new Error(`duplicate message parameter id: ${id.toString()}`);
				}
				const size = await r.u53();
				const bytes = await r.read(size);
				params.bytes.set(id, bytes);
			}
		}

		return params;
	}
}
